src/pybind/mgr/cephadm/upgrade.py

```diff
@@ -610,6 +603,14 @@ def _do_upgrade(self):
             for d in daemons:
                 if d.daemon_type != daemon_type:
                     continue
+                # if we don't have up-to-date metadata from this daemon's host we don't know
+                # for sure if the currently known image digest for this daemon is correct
+                # so we can't say for sure if it needs upgrade or if it's been upgraded (done)
+                if (
+                    self.mgr.use_agent
+                    and d.hostname in [h for h in self.mgr.cache.get_hosts() if not self.mgr.cache.host_metadata_up_to_date(h)]
+                ):
```

Comment:
```suggestion

    def agent_has_up_to_date_info(hostname: str): -> bool
        """
        if we don't have up-to-date metadata from this daemon's host we don't know
        for sure if the currently known image digest for this daemon is correct
        so we can't say for sure if it needs upgrade or if it's been upgraded (done)
        """
        return (
                    self.mgr.use_agent
                    and d.hostname in [h for h in self.mgr.cache.get_hosts() if not self.mgr.cache.host_metadata_up_to_date(h)]
                )
```

imo do_upgrade() is already complicated enough.

Reply:
realized I didn't need all this and we already have a function to check if a host's metadata is up to date in the host cache